### PR TITLE
Improve Person regrouping (EXPOSUREAPP-10154)

### PR DESCRIPTION
**Pull request by mtwalli**
## Ticket
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10154

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/storage/VaccinationStorage.kt [hunk 1]
```diff
@@ -56,23 +56,17 @@ class VaccinationStorage @Inject constructor(
...
                 remove(it)
             }
-            persons.forEach {
-                val raw = gson.toJson(it)
-                val identifier = it.identifier
-                Timber.tag(TAG).v("Storing vaccinatedPerson %s -> %s", identifier, raw)
-                putString("$PKEY_PERSON_PREFIX${identifier.groupingKey}", raw)
+            persons.groupDataByIdentifier().forEach {
+                if (it.vaccinations.isNotEmpty()) {
+                    val raw = gson.toJson(it)
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/storage/VaccinationStorage.kt:61]: checking for vaccinations as accessing identifier would crash if the are none

**Review by chiljamgossow — DISMISSED** [review]
Tested migration + update from 2.12

**Review by BMItr — DISMISSED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/storage/VaccinationStorage.kt [hunk 2]
```diff
@@ -56,45 +56,41 @@ class VaccinationStorage @Inject constructor(
...
-        save(load().groupDataByIdentifier())
-    }
-
     companion object {
         private const val TAG = "VaccinationStorage"
         private const val PKEY_PERSON_PREFIX = "vaccination.person."
     }
 }
 
 internal fun Set<VaccinatedPersonData>.groupDataByIdentifier(): Set<VaccinatedPersonData> {
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/storage/VaccinationStorage.kt:76]: for this fun the following **suggestion**:

internal fun Set<VaccinatedPersonData>.groupDataByIdentifier(): Set<VaccinatedPersonData> =
    filterNot { it.vaccinations.isNullOrEmpty() }
        .groupBy { it.identifier }
        .mapNotNull { entry ->
            val personDataList = entry.value
            if (personDataList.isEmpty()) {
                Timber.v("Person data list was empty, returning early")
                return@mapNotNull null
            }

            val newestData = personDataList.maxByOrNull {
                it.lastBoosterNotifiedAt ?: Instant.EPOCH
            }
            val vaccinations = personDataList.flatMap { it.vaccinations }.toSet()
            newestData?.copy(vaccinations = vaccinations)
        }.toSet()

**Review by BMItr — DISMISSED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/VaccinationRepository.kt [hunk 3]
```diff
@@ -73,7 +73,6 @@ class VaccinationRepository @Inject constructor(
     init {
         internalData.data
             .onStart { Timber.tag(TAG).d("Observing VaccinationContainer data.") }
-            .drop(1) // Initial emission, restored from storage.
             .onEach { vaccinatedPersons ->
                 Timber.tag(TAG).v("Vaccination data changed: %s", vaccinatedPersons)
                 storage.save(vaccinatedPersons.map { it.data }.toSet())
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/VaccinationRepository.kt:78]: the regrouping is only applied to the storage, not the value in the flow

**Review by BMItr — DISMISSED** [review]
image
badge tested successfully on oneplus 8
No crash could be spooted yet.

**Review by BMItr — DISMISSED** [review]

**Review by chiljamgossow — DISMISSED** [review]

**Review by BMItr — DISMISSED** [review]

**Review by chiljamgossow — DISMISSED** [review]

**Review by chiljamgossow — APPROVED** [review]

**Review by BMItr — APPROVED** [review]

**Comment by vaubaehn** [comment]
🤞

**Comment by vaubaehn** [comment]
According to Google Play Store Reviews the fix made many peolpe very happy 🎉 

Some community members and me are quite curious what actually happened.
As far as I understood the problem, it was a special case when applying booster notification rules on probably J&J jabs only?
And another community member wondered, how that influenced the amount of serialization that happened on start-up which is what the OOM stacktrace from the related issue points to?

Generally, we fully understand that the high time pressure likely made it impossible to document more on that hot fix. But we would be very happy, if you might return to the style some weeks/months ago, where PRs were documented with some more detail ❤️ 
Explaining PRs with "JIRA" or "TICKET" sometimes makes it hard to guess 😉 

Anyway, have a nice week and many greetings from all of us!

**Pull request merged**